### PR TITLE
sw: Add DOTP benchmark

### PR DESCRIPTION
**Pull request by mp-17**
And after all the noise around it, I share the implementation of DOTP that I am using.
Feel free to check it, fix it if needed, and add further optimizations. 

Let me know if there are any mistakes.

NOTE: as of now, the kernel only supports a single cluster.

### CODE REVIEW

**Comment by colluca** [comment]
Thanks @mp-17! I can take it on from here :)

File: sw/blas/dotp/scripts/datagen.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    def emit_header(self, **kwargs):
+        header = [super().emit_header()]
+
+        n = kwargs['n']
+        x = np.random.uniform(self.MIN, self.MAX, n)
+        y = np.random.uniform(self.MIN, self.MAX, n)
+        g = self.golden_model(x, y)
+
+        assert (n % 8) == 0, "n must be an integer multiple of the number of cores"
```
Comment (fischeti) [inline sw/blas/dotp/scripts/datagen.py:34]: For the unrolled kernel, each fraction of the vector also needs to be a multiple of 4 right?

File: sw/blas/dotp/src/main.c [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    snrt_cluster_hw_barrier();
+
+    if (!snrt_cluster_core_idx()) {
+      unsigned int runtime = end_cycle - start_cycle;
+      double performance   = (double) (2 * n - 1) / runtime;
+      double util          = 100 * (performance / (2 * snrt_cluster_compute_core_num()));
+
+      printf("Core %d execution time: %u cycles\nPerformance: %f DP-FLOP/Cycle\nUtilization: %f%%\n",
+        snrt_cluster_core_idx(), runtime, performance, util);
+    }
```
Comment (fischeti) [inline sw/blas/dotp/src/main.c:82]: Usually we extract performance metrics from the traces. But we report the utilization based on instructions and not FLOPs there.

File: sw/blas/dotp/scripts/datagen.py (hunk 1 above)
Comment (colluca) [reply sw/blas/dotp/scripts/datagen.py]: Good catch! I updated the script.

**Review by fischeti — APPROVED** [review]
LGTM. What would be nice is `fp8`, `fp16`, `fp32` support, but we can leave this for later.

**Review by colluca — APPROVED** [review]

**Pull request merged**